### PR TITLE
Implement mass cancel quotes (first pass)

### PR DESCRIPTION
**Pull request by jarry-xiao**
Here is an initial implementation of a mass cancel instruction. It's by no means rigorously tested yet, but I did add to the test file to test basic functionality. Note that if you remove either of the `mass_cancel` calls, the test will fail.

I don't know what the team's conventions are regarding style, but open to any and all feedback.

### CODE REVIEW

File: program/src/orderbook.rs [hunk 1]
```diff
@@ -307,6 +307,50 @@ impl<'ob> OrderBookState<'ob> {
...
+                    .to_vec();
+                if params
+                    .callback_id
+                    .iter()
+                    .zip(callback_info.iter())
+                    .all(|(a, b)| a == b)
+                {
+                    let order_id = leaf.order_id();
+                    book.remove_by_key(order_id);
+                    let cancel = Event::Out {
```
Comment (jarry-xiao) [inline program/src/orderbook.rs:334]: In my eyes the better interface here is to provide a buffer account that just contains a vector of order_id's that that can be passed back to the calling program so the quotes can all be processed synchronously. However, the event queue is the easiest thing to use as this doesn't break any existing interfaces

File: program/src/orderbook.rs [hunk 2]
```diff
@@ -307,6 +307,62 @@ impl<'ob> OrderBookState<'ob> {
         })
     }
 
+    pub fn mass_cancel(
+        &mut self,
+        params: &mass_cancel_quotes::Params,
+        event_queue: &mut EventQueue,
+    ) -> Result<OrderSummary, AoError> {
+        let mut num_orders_cancelled = 0;
+        let mut order_summary = OrderSummary {
```
Comment (jarry-xiao) [inline program/src/orderbook.rs:316]: Aggregate `total_base_qty` and `total_quote_qty` into the `OrderSummary` so the upstream program can track the amount left in book

**Comment by dr497** [comment]
Have you made any measurements to see what compute budget you save by adding this? I am afraid the overhead you save is pretty small

**Comment by jarry-xiao** [comment]
Haven’t made measurements yet, was just putting this out here as an idea. I’m guessing you would save at least the CPI cost (for programs that call into the AOB contract), but it is probably worth it to find the exact increase in throughput.

IMO it’s not so much about compute budget as wanting the mass cancel operation to be atomic. It’d be much easier to call this instruction than to call cancel in a loop in an upstream program or client. I think it also gets around potential tx size limitations.

Any other major opposition? I don’t think adding this hurts any existing functionality. There no urgency to merge it right now, but there might be future features that could benefit from having this instruction.

**Comment by jarry-xiao** [comment]
Update: just ran a quick test and verified that this instruction supports >200 new orders being cancelled at once. This is more than what you can do with `cancel_order` in 1 transaction.

It's not super likely that you'll have that many orders for a given user in the first place, but you could update the filter to just do substring matching and allow for a full orderbook clear. This could probably save a decent amount of work for protocols that want to repurpose/reclaim the rent of expired orderbooks.

**Comment by jarry-xiao** [comment]
I think you're right that it's not too much better as of right now. I was able to mass cancel 35 orders while using ~192k compute units. Not great.

Each leaf removal seems to take ~4k, and the initial in-order traversal took almost 60k! Maybe there's nothing much to gain here, but I think this is still probably 2-3x the number of cancels that you can fit into a normal instruction.

**Comment by dr497** [comment]
Can you describe a bit more how you would optimize the critbit?

**Comment by jarry-xiao** [comment]
Haven't thought it through 100%, but the potential idea would be:
1. Create a sorted doubly linked list in the LeafNode struct to allow for easier access. When you call `get_min` or `get_max` this will then return the head or tail of a linked list that can be traversed when handling `new_orders` or `mass_cancels`. Accessing the next node becomes a constant time lookup with this scheme. Reorganizing the inner nodes might be a little tricky here admittedly.
2. Potentially just mark nodes as `removed` instead of actually deleting them from the tree. You would need to have a some sort of scheme to reorder/cleanup.

These are somewhat half-baked thoughts, but I'll continue thinking about how to make some fo them more concrete.

**Comment by Henry-E** [comment]
Is the serum implementation of `prune` any use as a reference? https://github.com/project-serum/serum-dex/blob/master/dex/src/state.rs#L2567-L2610

**Comment by jarry-xiao** [comment]
> Is the serum implementation of `prune` any use as a reference? https://github.com/project-serum/serum-dex/blob/master/dex/src/state.rs#L2567-L2610
> 
> `prune` is essentially the same as what's being proposed here, remove all orders for a given user's open order account

This totally works with Serum-v3, but I don't think it will work here because there are no open orders accounts in the AOB. Orders/accounting are all managed by calling programs, so `prune` would have to be implemented in a higher level program. It would look roughly like this:

```
for order in user.orders {
    invoke_signed(
         &aaob_cancel_order(order, &cancel_order_accounts), &cancel_order_accounts, &[market_signer_seeds]
    )?;
}
```

This is undesirable due to the CPI overhead required. I think it would make a lot more sense to natively support the mass cancel operation on the tree level.

**Comment by Henry-E** [comment]
Ah, based on reading the serum v4 code it looked like the AOB was used as a crate rather than via CPI. Honestly, it seems like a really bad idea to use the AOB via CPI and not as a crate. It's adding unnecessary compute cost and an additional account. Efficiency is going to be really become even more important as solana changes starts changing its fee model and as more professional market makers are on boarded to the ecosystem.

**Comment by dr497** [comment]
The AOB is used as a lib in dex v4

**Comment by Henry-E** [comment]
Is a lib different from a crate?

**Pull request closed without merging**